bam_fastaref.c

```diff
@@ -0,0 +1,347 @@
+/*  bam_fastaref.c -- fastaref subcommand.
+
+    Copyright (C) 2017 Genome Research Ltd.
+
+    Author: Thomas Hickman <th10@sanger.ac.uk>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.  */
+
+#include "samtools.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <htslib/sam.h>
+#include <htslib/ref.h>
+#include <htslib/kstring.h>
+#include <htslib/hfile.h>
+
+/**
+ * Given a pointer to the start of a sam field, populates samField with a
+ * the malloced sam field.
+ * 
+ * Updates samFieldPosition to be the position after the sam field.
+ * 
+ * @returns 1 if successful
+*/
+int get_sam_field(char** samFieldPosition, char** samField){
+    char* samFieldStart = *samFieldPosition;
+    for(;**samFieldPosition != '\t' && **samFieldPosition != '\n' && **samFieldPosition != 0;(*samFieldPosition)++){};
+    int samFieldLength = *samFieldPosition - samFieldStart;
+
+    if((*samField = malloc(samFieldLength + 1)) == NULL) return 1;
+
+    strncpy(*samField, samFieldStart, samFieldLength);
+    (*samField)[samFieldLength] = '\0';
+
+    return 0;
+}
+
+typedef struct{
+    int maxLineLength;
+    char outputKeys[8][2];
+    int numOutputKeys;
+    char* samFileName;
+    char* outFileName; // NULL if this is to output to stdout
+} FastarefOptions;
+
+void init_kstring(kstring_t* str){
+    str->l = str->m = 0; str->s = NULL;
+}
+
+/**
+ * Given text of an SQ line beginning at sequenceText, populates values for the SN and M5
+ * fields and returns selected portions of the header according to options
+ * 
+ * This updates sequenceText after reading it.
+ */
+int getInfoForSQLine(char** sequenceText, char** SN, char** M5, kstring_t* restHeader, FastarefOptions *options){
```

Comment:
If you just want to parse the `@SQ` line, you should look into the functions `sam_hdr_parse` (`htslib/sam.c`), `sam_hdr_parse_` (`htslib/cram/sam_header.c`) and `bam_header_to_cram` (`htslib/cram/cram_samtools.c`). 
Unfortunately, we don't have an unified and homogeneous header API (it's high on our TO-DO list), but there is plenty of functionality scattered around, which you can use to implement your feature. There is no need to rewrite it.

Reply:
@valeriuo There doesn't seem to be any functions in `htslib/cram/sam_header.c` to get all headers which a specific tag (it looks like `sam_hdr_find` only returns the first header item). Would the expected thing to do be to expose the whole of the `SAM_hdr` data structure to `bam_fastaref.c`, or add a function to `sam_header.c` to expose this functionality?

Reply:
@ThomasHickman, check `sam_hdr_find_key` function (`sam_header.c`), with an example in `cram_io.c:1869`. Whatever is in `sam_header.h` is already exposed by HTSlib and you can use it in samtools.

Reply:
Sadly that function isn't exposed.  Only the htslib/*.h files are considered as external API.  The cram/*.h are internal and shouldn't be used outside htslib.  (Although in one case I did, out of sheer frustration!)

This is doubly annoying because we *do* have a header API, namely in the cram subdirectory, but not one blessed by htslib as OK to use.  This header API was derived from io_lib and arrived when I ported over CRAM, simply because I needed one for my own work.  We do really need to write down a list of requirements and then evaluate what we have to see if the current code fits it or whether we want to start afresh / amend it.